ansible-scylla-node/tasks/upgrade/pre_upgrade.yml

```diff
@@ -0,0 +1,165 @@
+  until: node_gossip_status.status == 200
+  retries: "{{ upgrade_api_default_retries |int }}"
+  delay: "{{ upgrade_api_default_delay |int }}"
+
+- name: Check if gossip is still enabled
+  ansible.builtin.fail:
+    msg: "Gossip it not disabled due to an unknown reason. Please check the node."
+  when: node_gossip_status.json |bool
+
+
+# Drain the node 
+- name: Drain the node
+  ansible.builtin.uri:
+    url: "http://{{ scylla_api_address }}:{{ scylla_api_port }}/storage_service/drain"
+    method: POST
+  register: node_drain_enable
+  until: node_drain_enable.status == 200
+  retries: "{{ upgrade_api_default_retries |int }}"
+  delay: "{{ upgrade_api_default_delay |int }}"
+
+- name: Ensure the node is fully drained
+  block:
+    - name: Ensure the node is fully drained
+      ansible.builtin.uri:
+        url: "http://{{ scylla_api_address }}:{{ scylla_api_port }}/storage_service/drain"
+        method: GET
+      register: node_drain_status
+      until: node_drain_status.json.find("Drained") != -1
+      retries: "{{ upgrade_api_default_retries |int * 3 }}"
+      delay: "{{ upgrade_api_default_delay |int }}"
+  rescue:
+    # Kill Scylla process
+    - name: Kill Scylla process
+      ansible.builtin.shell: killall --exact --signal SIGKILL scylla
+      register: scylla_process_kill
+      become: true
+    
+    # Enable snapshot generation
+    - name: Enable snapshot generation
+      ansible.builtin.debug:
+        msg: "Snapshot generation was set to be disabled but node draining failed. Enabling snapshot generation."
+      when: not upgrade_generate_snapshots
+
+
+# Prepare and create snapshot
+- name: Prepare and create snapshot
+  block:
+     - name: Create snapshot
+       ansible.builtin.uri:
+         url: "http://{{ scylla_api_address }}:{{ scylla_api_port }}/storage_service/snapshots?tag={{ scylla_upgrade['id'] }}"
+         method: POST
+       register: node_snapshot_create
+       until: node_snapshot_create.status == 200
+       retries: "{{ upgrade_api_default_retries |int }}"
+       delay: "{{ upgrade_api_default_delay |int }}"
+     
+     - name: Print snapshot identifier used
+       ansible.builtin.debug:
+         msg: "Snapshot identifier: {{ scylla_upgrade['id'] }}"
+  when: upgrade_generate_snapshots or scylla_process_kill is defined
+
+
+# Prepare and create backup
+- name: Create backup folders structure
+  ansible.builtin.file:
+    path : "{{ item }}"
+    state: directory
+    mode: '0755'
+  become: true
+  loop:
+    - "{{ upgrade_backup_path }}/{{ scylla_upgrade['id'] }}"
+    - "{{ upgrade_backup_path }}/{{ scylla_upgrade['id'] }}/etc"
+
+- name: Backup Scylla configuration file
+  ansible.builtin.copy:
+    src: /etc/scylla/scylla.yaml
+    dest: "{{ upgrade_backup_path }}/{{ scylla_upgrade['id'] }}/etc/scylla.yaml"
+    mode: preserve
+    remote_src: yes
+  become: true
+
+# ----- TODO: Check if there are other files to backup and how -----
+
+# Stop Scylla service
```

Comment:
just wanted to say kudoz for this .yml file
very nice fine grained stopping job :-)